lms/djangoapps/completion/models.py

```diff
@@ -141,45 +141,11 @@ def submit_batch_completion(self, user, course_key, blocks):
                 IntegrityError and OperationalError are relatively common
                 subclasses.
         """
-        if not waffle.waffle().is_enabled(waffle.ENABLE_COMPLETION_TRACKING):
-            raise RuntimeError(
-                "BlockCompletion.objects.submit_batch_completion should not be called when the feature is disabled."
-            )
-
-        block_key_to_block_completion = {}
+        block_completions = {}
         for block, completion in blocks:
-            block_key_to_block_completion[str(block)] = (block, completion)
-
-        block_to_was_created = {}
-        keys = [block for block, completion in blocks]
-
-        blocks_to_update = [block for block in BlockCompletion.objects.filter(
-            user=user,
-            course_key=course_key,
-            block_key__in=keys
-        )]
-
-        ignore_blocks = []
-        for block in blocks_to_update:
-            ignore_blocks.append(str(block.block_key))
-            block_key, completion = block_key_to_block_completion[str(block.block_key)]
-            block.completion = completion
-            block.save()
-            block_to_was_created[block] = False
-
-        for block_key in keys:
-            if str(block_key) in ignore_blocks:
-                continue
-            block_key, completion = block_key_to_block_completion[str(block_key)]
-            block = BlockCompletion(
-                user=user,
-                course_key=course_key,
-                block_key=block_key,
-                block_type=block_key.block_type,
-                completion=completion
-            )
-            block.save()
-        return block_to_was_created
+            (block_completion, isnew) = self.submit_completion(user, course_key, block, completion)
```

Comment:
The second part of [my previous comment](https://github.com/edx/edx-platform/pull/16688/commits/629c08f1c98a5c75a2b7215aec24cdd279fde2d4#r154279027) referred to using the existing `submit_completion` method.

I couldn't figure out why you needed all that other code, but if there was a reason, feel free to protest.

Reply:
Cliff provided this feedback the other week.
https://github.com/edx/edx-platform/pull/16688#discussion_r153546488

The idea was to create the least number of queries and allow this method to scale regardless of blocks. Testing the way you have it, with a basic test, seems to improve the querying from what I had, however, it will increase in size for every new block that is added. 

Each call to get_or_create results in at least 1 query, perhaps 2. What I did was limit it down to a reading/comparison query, and two sets of save queries, which hopefully, if atomic works as expected, should be 1 query to make 2 total queries.

Reply:
Letting you know that I didn't touched this yet until I heard back from you.